### PR TITLE
Clebsh-Gordan coefficients

### PR DESCRIPTION
**Pull request by L-sky**
1. Cache is tied to the same directory across executions now. Conventions on directory choice (across different operating systems) are uphold by appdirs library. 

2. Replaced svd with eigh to avoid unnecessary and unstable calculations for singular vectors corresponding to higher singular values. 

3. Combined wrapper functions - otherwise it is not possible to free memory of tensor A early, which makes a big difference in peak RAM usage.

Outputs agree with prior method up to machine level precision, except for some cases where sign may be flipped. Reason for that is sum of elements approaching zero, meaning that current sign-normalization is arbitrary (for both methods).

### CODE REVIEW

File: se3cnn/SO3.py [hunk 1]
```diff
@@ -478,32 +476,43 @@ def spherical_harmonics_coeff_to_sphere(coeff, alpha, beta):
...
+    ]
+
+    B = torch.zeros((n, n))                                                                             # preallocate memory
+    for abc in random_angles:                                                                           # expand block matrix multiplication with its transpose
+        D = _DxDxD(*abc) - torch.eye(n)
+        B += torch.matmul(D.t(), D)                                                                     # B = sum_i { D^T_i @ D_i }
+        del D
+        gc.collect()
+
+    s, v = scipy.linalg.eigh(B.numpy(), eigvals=(0, min(1, B.shape[0] - 1)), overwrite_a=True)          # ask for one (smallest) eigenvalue/eigenvector pair if there is only one exists, otherwise ask for two
```
Comment (mariogeiger) [inline se3cnn/SO3.py]: `B.shape[0] == n` ?

File: se3cnn/SO3.py [hunk 2]
```diff
@@ -478,32 +476,43 @@ def spherical_harmonics_coeff_to_sphere(coeff, alpha, beta):
...
+    n = (2 * l1 + 1) * (2 * l2 + 1) * (2 * l3 + 1)
+    random_angles = [
+        [4.41301023, 5.56684102, 4.59384642],
+        [4.93325116, 6.12697327, 4.14574096],
+        [0.53878964, 4.09050444, 5.36539036],
+        [2.16017393, 3.48835314, 5.55174441],
+        [2.52385107, 0.29089583, 3.90040975],
+    ]
+
+    B = torch.zeros((n, n))                                                                             # preallocate memory
```
Comment (mariogeiger) [inline se3cnn/SO3.py:503]: no need to preallocate => `B = 0`

File: se3cnn/SO3.py [hunk 3]
```diff
@@ -5,6 +5,7 @@
 Using ZYZ Euler angles parametrisation
 """
 import math
+from appdirs import user_cache_dir
```
Comment (mariogeiger) [inline se3cnn/SO3.py:8]: `appdirs` is not installed by default with anaconda. shall we add some `try: except:` that set `user_cache_dir` to the identity function in case `appdirs` is not installed?

File: se3cnn/SO3.py [hunk 4]
```diff
@@ -563,7 +571,7 @@ def clebsch_gordan(l1, l2, l3):
         return _clebsch_gordan(l3, l1, l2).transpose(0, 2).transpose(0, 1).contiguous()
 
 
-@cached_dirpklgz("cache/clebsch_gordan")
+@cached_dirpklgz(user_cache_dir("se3cnn/clebsch_gordan"))
 def _clebsch_gordan(l1, l2, l3, _version=2):
```
Comment (mariogeiger) [inline se3cnn/SO3.py]: change to `_version=3` if the output of the function may change

File: se3cnn/SO3.py [hunk 5]
```diff
@@ -478,32 +476,43 @@ def spherical_harmonics_coeff_to_sphere(coeff, alpha, beta):
...
-    # A = u @ torch.diag(s) @ v.t()
-    kernel = v.t()[s < eps]
-    return kernel
-
-
-def get_matrices_kernel(As, eps=1e-10):
-    """
-    Computes the commun kernel of all the As matrices
-    """
-    return get_matrix_kernel(torch.cat(As, dim=0), eps)
```
Comment (mariogeiger) [inline se3cnn/SO3.py:506]: I was concatenating the matrices together to get their null space (kernel is the same as null space).
I see that you are adding the matrices together instead.
I guess that the probability that you grow the null space by adding matrices together is small but still I prefer if we can make a more correct computation if we can.

for instance `[[1,0],[0,1]]` and `[[1,0],[0,-1]]` have no null space. But their sum has a dim 1 null space.

File: se3cnn/SO3.py [hunk 6]
```diff
@@ -478,32 +476,43 @@ def spherical_harmonics_coeff_to_sphere(coeff, alpha, beta):
...
-    :return: matrix where each row is a basis vector of the kernel of A
-    """
-    _u, s, v = torch.svd(A)
-
-    # A = u @ torch.diag(s) @ v.t()
-    kernel = v.t()[s < eps]
-    return kernel
-
-
-def get_matrices_kernel(As, eps=1e-10):
```
Comment (mariogeiger) [inline se3cnn/SO3.py:502]: I liked having dedicated function to get the null spaces of matrices.
I know I use them only once but they are well defined and we never know...

File: se3cnn/SO3.py [hunk 7]
```diff
@@ -478,32 +476,43 @@ def spherical_harmonics_coeff_to_sphere(coeff, alpha, beta):
...
-def get_matrix_kernel(A, eps=1e-10):
-    """
-    Compute an orthonormal basis of the kernel (x_1, x_2, ...)
-    A x_i = 0
-    scalar_product(x_i, x_j) = delta_ij
-
-    :param A: matrix
-    :return: matrix where each row is a basis vector of the kernel of A
-    """
-    _u, s, v = torch.svd(A)
```
Comment (mariogeiger) [inline se3cnn/SO3.py:495]: You should be able to improve this generic function for any null space dimension

**Review by mariogeiger — CHANGES_REQUESTED** [review]
Thanks for your PR Kostiantyn

File: se3cnn/SO3.py (hunk 1 above)
Comment (L-sky) [reply se3cnn/SO3.py]: true, that latter call to shape is excessive

File: se3cnn/SO3.py (hunk 5 above)
Comment (L-sky) [reply se3cnn/SO3.py:506]: concatenated `D` matrices form (block) matrix `A = [D1 D2 D3 D4 D5 ...]`, singular vectors of it are the same as eigenvectors of `B=A^T @ A`, as we have two compatible block matrices multiplied, we can expand expression block by block, which leads to the summation as in code

Comment (L-sky) [reply se3cnn/SO3.py]: for your example it will be 
```[[1, 0], [0, 1]]^T @ [[1, 0], [0, 1]] + [[1, 0], [0, -1]]^T @ [[1, 0], [0, -1]] = [[1, 0], [0, 1]] + [[1, 0], [1, 0]] = 2*[[1, 0], [0, 1]]
``` - no null space

Comment (L-sky) [reply se3cnn/SO3.py:506]: for your example it will be 
`
[[1,0],[0,1]]^T @ [[1,0],[0,1]] + [[1,0],[0,-1]]^T @ [[1,0],[0,-1]] = 2 * [[1,0],[0,1]]
`
no null space

File: se3cnn/SO3.py (hunk 2 above)
Comment (L-sky) [reply se3cnn/SO3.py:503]: while technically it is true that first `+=` is going to reassign value with such initialization, I am a little bit cautious about impact on ref count in `del`

File: se3cnn/SO3.py (hunk 3 above)
Comment (L-sky) [reply se3cnn/SO3.py:8]: add to requirements, maybe? it is 12 kB standalone, and saves a lot of headache

File: se3cnn/SO3.py (hunk 4 above)
Comment (L-sky) [reply se3cnn/SO3.py]: incremented

File: se3cnn/SO3.py (hunk 5 above)
Comment (mariogeiger) [reply se3cnn/SO3.py:506]: clever, could we put this trick into `get_matrices_kernel` ?

File: se3cnn/SO3.py (hunk 3 above)
Comment (mariogeiger) [reply se3cnn/SO3.py:8]: ok

File: se3cnn/SO3.py (hunk 2 above)
Comment (mariogeiger) [reply se3cnn/SO3.py:503]: as you wish

File: se3cnn/SO3.py (hunk 5 above)
Comment (L-sky) [reply se3cnn/SO3.py:506]: not really, as it uses svd we never have access to `A^T @ A` even if internals construct it (which is not necessarily happens, depends on underlying algorithm)

File: se3cnn/SO3.py (hunk 7 above)
Comment (L-sky) [reply se3cnn/SO3.py:495]: I am a bit confused, can you expand on this one?

File: se3cnn/SO3.py (hunk 6 above)
Comment (L-sky) [reply se3cnn/SO3.py:502]: it is not like I much into erasing, but do we really need 
`_u, s, v = torch.svd(A); return v.t()[s < eps]` 
which is no longer referenced anywhere, to be kept in active version of the code, rather than just in history

File: se3cnn/SO3.py (hunk 5 above)
Comment (mariogeiger) [reply se3cnn/SO3.py:506]: I don't understand the problem.
`get_matrices_kernel(Ds)` gets a list of matrices. it just has to return `get_matrix_kernel(D1.T @ D1 + D2.T @ D2 + D3.T @ D3)`

Comment (L-sky) [reply se3cnn/SO3.py:506]: if we do so, then we have to use eigh or similar,
and if you mean why not keep original function names/structure, while changing only few lines inside, the reason is that in python one cannot free memory of an object passes to the function within said function, and that difference appeared quite substantial.

Comment (mariogeiger) [reply se3cnn/SO3.py:506]: yes, let use eigh or similar!
Optimization makes the code more ugly :-( 
Do we really need this memory ?^^

File: se3cnn/SO3.py (hunk 6 above)
Comment (mariogeiger) [reply se3cnn/SO3.py:502]: Of course if you don't call it at all, remove it.
My suggestion was to modify it and call it once. But now I understand your concern about the garbage collector and its interaction with function arguments.

File: se3cnn/SO3.py (hunk 7 above)
Comment (mariogeiger) [reply se3cnn/SO3.py:495]: I am wondering if your use of `eigh` is suitable for null spaces of any dimensions

File: se3cnn/SO3.py (hunk 6 above)
Comment (L-sky) [reply se3cnn/SO3.py:502]: also, about nullspace, it may be not the best idea to use pytorch svd for that in general, as it targets largest singular values. 
In contrast, there is `scipy.sparse.linalg.svds` (with 'SM' mode). Despite `sparse` in name it actually accepts dense arrays.    
https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.svds.html#scipy.sparse.linalg.svds

File: se3cnn/SO3.py (hunk 7 above)
Comment (L-sky) [reply se3cnn/SO3.py:495]: it works for any, though you will receive not all if asked number of vectors is smaller than null space

File: se3cnn/SO3.py (hunk 5 above)
Comment (L-sky) [reply se3cnn/SO3.py]: I personally do not, because I do not use higher order CG, 
but it makes a difference whether one can fully calculate L=9+ within 16+2 GB of RAM

Comment (L-sky) [reply se3cnn/SO3.py:506]: it makes difference whether one can fully calculate L=9 and higher within 16+2 GB of RAM

File: se3cnn/SO3.py (hunk 6 above)
Comment (mariogeiger) [reply se3cnn/SO3.py:502]: Go ahead with any function better than `torch.svd`.
By looking at the `scipy.sparse.linalg.svds` documentation I saw a note:
> This is a naive implementation using ARPACK as an eigensolver on A.H * A or A * A.H, depending on which one is more efficient.

I worrying that making several times `A.T @ A` will add numerical instabilities.

File: se3cnn/SO3.py (hunk 5 above)
Comment (mariogeiger) [reply se3cnn/SO3.py:506]: ok then

File: se3cnn/SO3.py (hunk 7 above)
Comment (mariogeiger) [reply se3cnn/SO3.py:495]: That was my concern. Then I guess we get rid of `get_matrix_kernel` since there is now 2 reasons to not keep it.

File: se3cnn/SO3.py (hunk 6 above)
Comment (L-sky) [reply se3cnn/SO3.py:502]: why several, just once 

Anyway, as for the task in hand, choice was between `eigh` and `eigsh` due to tighter memory control, and former slightly nicer, because of additional options

File: se3cnn/SO3.py (hunk 3 above)
Comment (L-sky) [reply se3cnn/SO3.py:8]: added
